### PR TITLE
Add wasm gov proposals to cli

### LINKED ISSUE #178: Gov: cli support
## Summary

Provide CLI support for new proposal types in #162 

Relates to #161
____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### LINKED ISSUE #184: CLI: gov store proposal wasm code display issue
## Problem Definition
When proposals listed or a single one is displayed via CLI the wasm byte code is rendered byte by byte on a new line. This Is due to the default YAML rendering in `cliCtx.PrintOutput()`.

```yaml
- content:
    wasmproposal:
      title: foo
      description: bar
    creator: cosmos100dejzacpanrldpjjwksjm62shqhyss44jf5xz
    wasm_byte_code:
    - 31
    - 139
    - 8
    - 0
    - 0
    - 0
    - 0
    - 0
    - 0
    - 255
    - 236
    - 189
    - 13
    - 148
    - 30
    - 199
    - 113
    - 24
    - 216
    - 213
    - 221
    - 51
    - 223
    - 183
    - 223
    - 204
    - 236
    - 14
...
```

This can be reproduced with
```
# store
wasmcli tx gov submit-proposal wasm-store contract.wasm --title=foo --description=bar --creator=cosmos100dejzacpanrldpjjwksjm62shqhyss44jf5xz --from validator --gas 10000000  -y --chain-id=testing --node=http://localhost:26657 -b block

# list proposals
wasmcli q gov proposals
```

### Workaround
Use json output via `--output json` flag.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by alpe**
Resolves #178
Resolves #184

### New commands
```
Usage:
  wasmcli tx gov submit-proposal [flags]
  wasmcli tx gov submit-proposal [command]

Available Commands:
  wasm-store           Submit a wasm binary proposal
  instantiate-contract Submit an instantiate wasm contract proposal
  migrate-contract     Submit a migrate wasm contract to a new code version proposal
  set-contract-admin   Submit a new admin for a contract proposal
  clear-contract-admin Submit a clear admin for a contract to prevent further migrations proposal
```

## Disclaimer
I have manually tested all commands and submitted gov proposals via CLI. 

Still open to test
- [ ] The JSON parsers that convert json to proposals
- [ ] Proposal execution

### Store

```
  wasmcli tx gov submit-proposal wasm-store [wasm file] --source [source] --builder [builder] --title [text] --description [text] --creator [address] [flags]
```
Example:
```
wasmcli tx gov submit-proposal wasm-store contract.wasm --title=foo --description=bar --creator=cosmos100dejzacpanrldpjjwksjm62shqhyss44jf5xz --from validator --gas 10000000  -y --chain-id=testing --node=http://localhost:26657 -b block
```
⚠️ The wasm byte code creates UI issues when listing proposals later. See #184  ⚠️ 

### Instantiate
```
wasmcli tx gov submit-proposal instantiate-contract [code_id_int64] [json_encoded_init_args] --label [text] --admin [address] --title [text] --description [text] --creator [address] [flags]
```
Example
```
wasmcli tx gov submit-proposal instantiate-contract 1 '{"verifier":"cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8","beneficiary":"cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8"}' --label my-test --admin cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8 --title instantiate --description test --creator cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8 --from validator --gas 10000000   --chain-id=testing --node=http://localhost:26657 -b block
```
### Migrate
```sh
wasmcli tx gov submit-proposal migrate-contract [contract_addr_bech32] [new_code_id_int64] [json_encoded_migration_args] [flags]
```
Example
```sh
wasmcli tx gov submit-proposal migrate-contract cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5 2 '{"verifier": "cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8"}' --title instantiate --description test --sender cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8 --from validator --gas 10000000   --chain-id=testing --node=http://localhost:26657 -b block
```

### Update Admin

```
wasmcli tx gov submit-proposal set-contract-admin [contract_addr_bech32] [new_admin_addr_bech32] [flags]
```
Example
```
wasmcli tx gov submit-proposal set-contract-admin cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5 cosmos1rsj2m3qas3jy50s8psns95qka09lhwhx2wvdq8 --title set-admin --description test  --from validator --gas 10000000   --chain-id=testing --node=http://localhost:26657 -b block
```

### Clear Admin
```
wasmcli tx gov submit-proposal clear-contract-admin [contract_addr_bech32] [flags]
```
Example
```
wasmcli tx gov submit-proposal clear-contract-admin cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5  --title clear-admin --description test  --from validator --gas 10000000   --chain-id=testing --node=http://localhost:26657 -b block
```

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by ethanfrey** [comment]
Thank you for the examples. I will look at the code, but the cli commands look good. The only question I have is the

`--sender` flag on migrate and `--creator` on store and instantiate

Maybe they can be the same name? Maybe a different name?

File: cmd/wasmcli/main.go [hunk 1]
```diff
@@ -127,12 +127,17 @@ func txCmd(cdc *amino.Codec) *cobra.Command {
 	// add modules' tx commands
 	app.ModuleBasics.AddTxCommands(txCmd, cdc)
 
+	//wasmGovCmds := []*cobra.Command{}
```
Comment (ethanfrey) [inline cmd/wasmcli/main.go:130]: ????

File: x/wasm/client/cli/gov_tx.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+				return err
+			}
+			if len(viper.GetString(flagCreator)) == 0 {
+				return errors.New("creator address is required")
+			}
+			creator, err := sdk.AccAddressFromBech32(viper.GetString(flagCreator))
+			if err != nil {
+				return errors.Wrap(err, "creator")
+			}
+			content := types.StoreCodeProposal{
```
Comment (ethanfrey) [inline x/wasm/client/cli/gov_tx.go:41]: Looks good

File: x/wasm/client/cli/tx.go [hunk 3]
```diff
@@ -103,58 +82,70 @@ func StoreCodeCmd(cdc *codec.Codec) *cobra.Command {
 
 	cmd.Flags().String(flagSource, "", "A valid URI reference to the contract's source code, optional")
 	cmd.Flags().String(flagBuilder, "", "A valid docker tag for the build system, optional")
+	cmd.Flags().String(flagInstantiateByEverybody, "", "Everybody can instantiate a contract from the code, optional")
```
Comment (ethanfrey) [inline x/wasm/client/cli/tx.go:85]: I like these flags.

**Comment by alpe** [comment]
> `--sender` flag on migrate and `--creator` on store and instantiate

🤔   I don't have a good idea for better naming but this is some context for the different roles:

* `Store.Creator` is the address that "owns" the code object.
* `Instantiate.Creator` is the address that pays the init funds. It is the creator of the contract and passed to the contract as `sender`
* `Migrate.Sender`  is the address that is passed to the contract's environment. Should be admin but for gov can be any address.

File: cmd/wasmcli/main.go (hunk 1 above)
Comment (alpe) [reply cmd/wasmcli/main.go]: removed

**Comment by ethanfrey** [comment]
> I don't have a good idea for better naming but this is some context for the different roles:

Yeah, I saw that later. I guess my approach reading the samples was to seek to make them all common, but you are right, they have different meanings and you follow the name of the fields.

My issue is that the word `--creator` is a bit confusing in the context of a proposal. The one creating the proposal?

Maybe the word `--run-as` to say the account the passed proposal will execute as. Or maybe you have a better idea.

This is a minor point, we can leave it as is. But I doubt I will look at this for quite some time, so good to clarify the naming now if possible. If you like `--run-as` or have another idea, go for it. Otherwise, we can leave it as you have it.

**Comment by alpe** [comment]
`--run-as` sounds better. I will apply the changes

**Comment by ethanfrey** [comment]
Looks good. And nice to add the `--amount` flag to the docs as well, I missed that one.

From my side, feel free to merge

**Pull request merged**